spinalcordtoolbox/labels.py

```diff
@@ -464,3 +465,69 @@ def remove_other_labels_from_image(img: Image, labels: Sequence[int]) -> Image:
             logger.warning(f"Label {label} not found in input image!")
 
     return out
+
+
+def project_discs(img: Image, ref: Image, num_disc=True) -> Image:
+    """
+    Project discs coordinates on the spinal cord centerline. This projection is obtained by iterating along
+    the centerline to identify the shortest distance with each referenced coordinates.
+    Typically, user inputs a segmentation image, and labels with disks position, and this function computes
+    the identification to the closest coordinates of each labels on the centerline.
+    Labels are assumed to be non-zero and incremented from top to bottom
+
+    :param img: segmentation
+    :param ref: reference labels
+    :param num_disc: If True, projected disc labels have the value of the label, otherwise the value is set to 1.
+    :returns: image with the new projected discs labels
+    """
+    # Checking orientation
+    if img.orientation != "RPI":
+        img.change_orientation("RPI")
+
+    if ref.orientation != "RPI":
+        ref.change_orientation("RPI")
+
+    # Extract centerline from segmentation
+    _, arr_ctl, _, _ = get_centerline(img, param=ParamCenterline())
+    centerline = arr_ctl.T
+
+    # Extract referenced coordinates
+    coordinates_ref = ref.getNonZeroCoordinates(sorting='value')
+
+    # Create function to compute the projection on the centerline
+    def projection(point, num_disc=False):
+        """
+        Project the input point by minimizing its distance with the centerline
+        """
+        # Separate disc number from coordinates
+        if num_disc:
+            disc = point[-1]
+            point = point[:3]
+
+        # Define distance function between the point and the centerline
+        def distance(t):
+            return np.linalg.norm(np.array(point) - np.array(centerline[int(t)]))
+
+        # Minimize distance function to find the parameter t of the projection
+        result = minimize_scalar(distance, bounds=(0, centerline.shape[0]-1), method='bounded')
```

Comment:
I don't think `minimize_scalar` is the right approach: it's designed for finding the minimum of a continuous function on a real interval, but what you really have is a discontinuous function on a discrete interval (really, a finite list of points). It would be better to just iterate over the centerline (as you wrote in the documentation, actually!).

You can either write out the double loop manually (outer loop over the labelled points, inner loop over the centerline points), or there are a few ways to replace the inner loop with a call to [the builtin `min` function](https://docs.python.org/3/library/functions.html#min) (by taking the minimum of a list of (distance, index) pairs, or by using the `key=` parameter of `min`).

Reply:
I generally avoid creating nested loops to improve computation time but if you feel it is the right approach, I can change that.